compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -449,11 +451,22 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                     }
                 }
             }
+            ExprKind::AddressOf { .. } => {
+                // we want to forgive one deref, so addr_of!(STATIC_MUT) works
+                self.in_addr_of = true;
+            }
             ExprKind::Deref { arg } => {
+                // during HIR -> THIR, we synthesize a pointer to the static and then deref it
+                let allow_implicit_static_deref = self.in_addr_of;
+                // we don't want to accidentally allow addr_of!(*STATIC_MUT)
```

Comment:
Do we have a test ensuring that we do not allow that?

Reply:
I think one of the "static-with-pointer-to-static" tests does this, but I'll double-check.